Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -326,35 +327,98 @@ PhysicalParticleContainer::AddGaussianBeam (
+    //TODO: Add ASSERT_WITH_MESSAGE to test if mass and charge are both const
+    amrex::ParticleReal p_m = ps.second["mass"][openPMD::RecordComponent::SCALAR].loadChunk<amrex::ParticleReal>().get()[0];
+    double const mass_unit = ps.second["mass"][openPMD::RecordComponent::SCALAR].unitSI();
+    amrex::ParticleReal p_q = ps.second["charge"][openPMD::RecordComponent::SCALAR].loadChunk<amrex::ParticleReal>().get()[0];
+    double const charge_unit = ps.second["charge"][openPMD::RecordComponent::SCALAR].unitSI();
+#if (defined WARPX_DIM_3D) || (defined WARPX_DIM_2D)
+    auto const npart = ps.second["position"]["x"].getExtent()[0];
+    std::shared_ptr<amrex::ParticleReal> ptr_x = ps.second["position"]["x"].loadChunk<amrex::ParticleReal>();
+    double const position_unit_x = ps.second["position"]["x"].unitSI();
+    std::shared_ptr<amrex::ParticleReal> ptr_z = ps.second["position"]["z"].loadChunk<amrex::ParticleReal>();
+    double const position_unit_z = ps.second["position"]["z"].unitSI();
+    std::shared_ptr<amrex::ParticleReal> ptr_ux = ps.second["momentum"]["x"].loadChunk<amrex::ParticleReal>();
+    double const momentum_unit_x = ps.second["momentum"]["x"].unitSI();
+    std::shared_ptr<amrex::ParticleReal> ptr_uz = ps.second["momentum"]["z"].loadChunk<amrex::ParticleReal>();
+    double const momentum_unit_z = ps.second["momentum"]["z"].unitSI();
+#else
+    amrex::Abort("AddPlasmaFromFile is only implemented for 2D and 3D\n")
+#endif
+#if (defined WARPX_DIM_3D)
+    std::shared_ptr<amrex::ParticleReal> ptr_y = ps.second["position"]["y"].loadChunk<amrex::ParticleReal>();
+    double const position_unit_y = ps.second["position"]["y"].unitSI();
+    std::shared_ptr<amrex::ParticleReal> ptr_uy = ps.second["momentum"]["y"].loadChunk<amrex::ParticleReal>();
+    double const momentum_unit_y = ps.second["momentum"]["y"].unitSI();
+#endif
     series.flush();
 
-    mass = p_m*PhysConst::mevpc2_kg;
-    charge = p_q*PhysConst::q_e;
-    Real const weight = q_tot/(charge*amrex::Real(npart));
+    mass=p_m*mass_unit;
+    charge=p_q*charge_unit;
 
-    amrex::Print() << npart << " parts of species " << ps.first << "\nWith"
-    << " mass = " << mass << " and charge = " << charge << "\nTo initialize "
-    << npart << " macroparticles with weight " << weight << "\n";
+    amrex::Real weight;
+    if (q_tot != 0.0){
+        weight = q_tot/(p_q*amrex::Real(npart));
+    }
+    else {
+        weight = charge;
+    }
+
+    // Declare temporary vectors on the CPU
+    Gpu::HostVector<ParticleReal> particle_x;
+    Gpu::HostVector<ParticleReal> particle_z;
+    Gpu::HostVector<ParticleReal> particle_ux;
+    Gpu::HostVector<ParticleReal> particle_uz;
+    Gpu::HostVector<ParticleReal> particle_w;
+    Gpu::HostVector<ParticleReal> particle_y;
+    Gpu::HostVector<ParticleReal> particle_uy;
 
-    amrex::Print()<<"WARNING: this is WIP, no particle has been injected!!";
+    if (ParallelDescriptor::IOProcessor()) {
```

Comment:
this needs to move up to the beginning of the function body before opening the series :)